File: solutions/problem_2146_k_highest_ranked_items_within_a_price_range/solution.cpp
/*
 * Problem 2146: K Highest Ranked Items Within a Price Range
 * =======================================================
 * Difficulty: Medium
 * Tags: Array, Breadth-First Search, Sorting, Heap (Priority Queue), Matrix
 * Pattern: Heap / Priority Queue
 *
 * Time Complexity:  O(n log n)
 * Space Complexity: O(n)
 */

#include <queue>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> highestRankedKItems(vector<vector<int>>& grid, vector<int>& pricing, vector<int>& start, int k) {
        // Heap/Priority Queue - O(n log k) time
        priority_queue<int, vector<int>, greater<int>> pq;
        for (int val : grid) {
            pq.push(val);
            if ((int)pq.size() > pricing)
                pq.pop();
        }
        return pq.empty() ? {} : pq.top();
    }
};

File: solutions/problem_2146_k_highest_ranked_items_within_a_price_range/solution.py
"""
Problem 2146: K Highest Ranked Items Within a Price Range
=======================================================
Difficulty: Medium
Tags: Array, Breadth-First Search, Sorting, Heap (Priority Queue), Matrix
Pattern: Heap / Priority Queue

Time Complexity:  O(n log n)
Space Complexity: O(n)
"""

class Solution:
    def highestRankedKItems(self, grid: List[List[int]], pricing: List[int], start: List[int], k: int) -> List[List[int]]:
        # Heap/Priority Queue - O(n log k) time
        import heapq
        if not grid:
            return []
        # Min heap (negate for max heap)
        heap = []
        for val in grid:
            heapq.heappush(heap, val)
            if len(heap) > (pricing if isinstance(pricing, int) else len(grid)):
                heapq.heappop(heap)
        return heap[0] if heap else []
